### PR TITLE
Fix DiskReadViolation warning when opening cached Realms.

### LINKED ISSUE #7500: RealmResults.toFlow() extension does disk I/O
#### Goal
We are migrating our project to using Flow in our codebase (we used to use LiveData).

Thus we call `toFlow` on our `RealmResults`.

#### Actual Results
When doing so, we notice StrictMode complaining about this method reading from the disk - and since we are calling this from our ViewModel's init block (which I think is fairly reasonable), we are in the main thread at that point.

Here's the StrictMode stacktrace:

```
2021-06-15 11:51:25.085 12924-12924/eu.qonto.qonto.debug D/StrictMode: StrictMode policy violation; ~duration=19 ms: android.os.strictmode.DiskReadViolation
        at android.os.StrictMode$AndroidBlockGuardPolicy.onReadFromDisk(StrictMode.java:1596)
        at libcore.io.BlockGuardOs.access(BlockGuardOs.java:71)
        at libcore.io.ForwardingOs.access(ForwardingOs.java:72)
        at android.app.ActivityThread$AndroidOs.access(ActivityThread.java:7533)
        at java.io.UnixFileSystem.checkAccess(UnixFileSystem.java:281)
        at java.io.File.exists(File.java:815)
        at io.realm.RealmConfiguration.realmExists(RealmConfiguration.java:257)
        at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:423)
        at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:412)
        at io.realm.Realm.getInstance(Realm.java:424)
        at io.realm.internal.coroutines.InternalFlowFactory$from$3.invokeSuspend(InternalFlowFactory.kt:122)
        at io.realm.internal.coroutines.InternalFlowFactory$from$3.invoke(Unknown Source:10)
        at kotlinx.coroutines.flow.ChannelFlowBuilder.collectTo$suspendImpl(Builders.kt:341)
        at kotlinx.coroutines.flow.ChannelFlowBuilder.collectTo(Unknown Source:0)
        at kotlinx.coroutines.flow.CallbackFlowBuilder.collectTo(Builders.kt:355)
        at kotlinx.coroutines.flow.internal.ChannelFlow$collectToFun$1.invokeSuspend(ChannelFlow.kt:60)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.EventLoop.processUnconfinedEvent(EventLoop.common.kt:69)
        at kotlinx.coroutines.internal.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:385)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:30)
        at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable$default(Cancellable.kt:25)
        at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:110)
        at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
        at kotlinx.coroutines.BuildersKt.launch(Unknown Source:1)
        at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
        at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source:1)
        at androidx.lifecycle.BlockRunner.maybeRun(CoroutineLiveData.kt:174)
        at androidx.lifecycle.CoroutineLiveData.onActive(CoroutineLiveData.kt:240)
        at androidx.lifecycle.LiveData.changeActiveCounter(LiveData.java:390)
        at androidx.lifecycle.LiveData$ObserverWrapper.activeStateChanged(LiveData.java:466)
        at androidx.lifecycle.LiveData$LifecycleBoundObserver.onStateChanged(LiveData.java:425)
        at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:354)
        at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:265)
        at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:307)
        at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.java:148)
        at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.java:134)
        at androidx.lifecycle.ReportFragment.dispatch(ReportFragment.java:68)
        at androidx.lifecycle.ReportFragment$LifecycleCallbacks.onActivityPostStarted(ReportFragment.java:187)
        at android.app.Activity.dispatchActivityPostStarted(Activity.java:1362)
        at android.app.Activity.performStart(Activity.java:8061)
        at android.app.ActivityThread.handleStartActivity(ActivityThread.java:3475)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:221)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

When digging a bit, we can see that this is provoked by calling `Realm.getInstance(config)` line 122 of `InternalFlowFactory`.

#### Steps & Code to Reproduce
Call `toFlow` from the main thread, with StrictMode enabled.

#### Version of Realm and tooling

Realm version(s): 10.5.0

Realm Sync feature enabled: No

Android Studio version: 4.2.1

Android Build Tools version: not sure / the latest

Gradle version: 7.0.2

Which Android version and device(s): sdk 30 (emulator).

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #7500 

Verified locally that this works:

```
var realm = Realm.getDefaultInstance()

        StrictMode.setThreadPolicy(
            StrictMode.ThreadPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .penaltyDeath()
                .build()
        )

realm = Realm.getDefaultInstance()
```

### CODE REVIEW

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java [hunk 1]
```diff
@@ -4792,4 +4793,38 @@ public void getNumberOfActiveVersions() throws InterruptedException {
...
+    @Test
+    public void getCachedInstanceDoNotTriggerStrictMode() {
+        StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
+            .detectAll()
+            .penaltyLog()
+            .penaltyDeath()
+            .build());
+        try {
+            Realm.getInstance(realmConfig).close();
+        } finally {
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:4806]: isn't this swallowing the exception thrown by `getInstance`?

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java [hunk 2]
```diff
@@ -4792,4 +4793,38 @@ public void getNumberOfActiveVersions() throws InterruptedException {
...
+
+    @Test
+    public void getCachedInstanceFromOtherThreadDoNotTriggerStrictMode() throws InterruptedException {
+        StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
+                .detectAll()
+                .penaltyLog()
+                .penaltyDeath()
+                .build());
+        try {
+            Thread t = new Thread(() -> Realm.getInstance(realmConfig).close());
```
Comment (nhachicha) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:4821]: should you capture potential exception thrown in thread `t`?

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java (hunk 1 above)
Comment (cmelchior) [reply realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:4806]: No, it just runs whether or not an exception is thrown.

File: realm/realm-library/src/androidTest/java/io/realm/RealmTests.java (hunk 2 above)
Comment (cmelchior) [reply realm/realm-library/src/androidTest/java/io/realm/RealmTests.java:4821]: There shouldn't be a need. An exception in this thread will correctly fail the test.

**Pull request merged**